src/Peachpie.CodeAnalysis/Utilities/PriorityQueue.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+namespace Peachpie.CodeAnalysis.Utilities
+{
+    /// <summary>
+    /// PriorityQueue provides a stack-like interface, except that objects
+    /// "pushed" in arbitrary order are "popped" in order of priority, i.e., 
+    /// from least to greatest as defined by the specified comparer.
+    /// </summary>
+    /// <remarks>
+    /// Push and Pop are each O(log N). Pushing N objects and them popping
+    /// them all is equivalent to performing a heap sort and is O(N log N).
+    /// Multiple different items of the same priority are allowed, but their
+    /// order is not guaranteed to be stable.
+    /// 
+    /// The original implementation was taken from the Microsoft .NET
+    /// Framework Reference Source, author: Niklas Borson (niklasb).
+    /// </remarks>
+    internal class PriorityQueue<T>
```

Comment:
found issue in Microsoft impl - https://stackoverflow.com/questions/44221454/bug-in-microsofts-internal-priorityqueuet

Do we fix this?

Reply:
Wow.. I should've checked that. Anyway, it's fixed now in b34ed05.